### PR TITLE
Parallel replicas always skip unavailable ones

### LINKED ISSUE #48565: Parallel replicas should skip unavailable hosts
https://pastila.nl/?0438131e/2cec234307914a8796d2283cd7a33894

See also #46314

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
When parallel replicas are enabled they will always skip unavailable servers (the behavior is controlled by the setting `skip_unavailable_shards`, enabled by default and can be only disabled). This closes: https://github.com/ClickHouse/ClickHouse/issues/48565

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 43b8513bc19093bae3f5b37418b177aa0b4bb3a1 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50293/43b8513bc19093bae3f5b37418b177aa0b4bb3a1/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by qoega** [comment]
Why shards but not replicas in the name?

**Comment by alexey-milovidov** [comment]
Because parallel replicas are dynamic shards.

But we also should take care about consistency with the implementation of the same behavior in the s3Cluster and urlCluster table functions.

**Comment by nikitamikhaylov** [comment]
Looks like it already works for `*Cluster` table functions https://github.com/ClickHouse/ClickHouse/blob/688f23b32571a253ae532fc8c5655610bef022e7/src/Storages/IStorageCluster.cpp#L137

**Comment by alexey-milovidov** [comment]
Yes, see https://github.com/ClickHouse/ClickHouse/issues/46314
My point is that the implementation has to be consistent.

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 1]
```diff
@@ -82,6 +87,7 @@ class ParallelReplicasReadingCoordinator::ImplInterface
     virtual ~ImplInterface() = default;
     virtual ParallelReadResponse handleRequest(ParallelReadRequest request) = 0;
     virtual void handleInitialAllRangesAnnouncement(InitialAllRangesAnnouncement announcement) = 0;
+    virtual void markReplicaAsUnavailble(size_t replica_number) = 0;
```
Comment (serxa) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:90]: typo
```suggestion
    virtual void markReplicaAsUnavailable(size_t replica_number) = 0;
```

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 2]
```diff
@@ -512,6 +540,18 @@ void ParallelReplicasReadingCoordinator::setMode(CoordinationMode mode_)
     mode = mode_;
```
Comment (serxa) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:540]: Probably `setMode()` method was intended to be private? It does not create `pimpl`

**Review by serxa — APPROVED** [review]

**Pull request merged**